[flang][NFC] use llvm.intr.stacksave/restore instead of opaque calls

tblah opened the pull request:
The new LLVM stack save/restore intrinsic operations are more convenient than function calls because they do not add function declarations to the module and therefore do not block the parallelisation of passes. Furthermore they could be much more easily marked with memory effects than function calls if that ever proved useful.

This builds on top of #107879.

Resolves #108016

tblah committed:
[flang][NFC] use llvm.intr.stacksave/restore instead of opaque calls

The new LLVM stack save/restore intrinsic operations are more convenient
than function calls because they do not add function declarations to the
module and therefore do not block the parallelisation of passes.
Furthermore they could be much more easily marked with memory effects
than function calls if that ever proved useful.

This builds on top of #107879.

Resolves #108016

clementval reviewed (approved):
LGTM

jeanPerier commented on flang/lib/Optimizer/Builder/FIRBuilder.cpp:
```diff
@@ -328,6 +329,18 @@ mlir::Value fir::FirOpBuilder::createHeapTemporary(
 }
 
+mlir::Value fir::FirOpBuilder::genStackSave(mlir::Location loc) {
+  mlir::DataLayout dataLayout(getModule());
```
I have mixed feeling about having the mlir::DataLayout be built here, mostly because MLIR docs suggest the data layout computation may be "expensive" here: 

https://github.com/llvm/llvm-project/blob/af1d2b9fb1c6065d3a4611f086b8c83a0c05795c/mlir/docs/DataLayout.md?plain=1#L84

`genStackSave` is not as hot as some helpers to get types, so it is likely acceptable here, but I am worried people may just copy paste this solution to get a DataLayout here and there. I think I would make that unique_ptr member of the fir::FirOpBuilder built on demand in some getDataLayout() accessor if not already built.

jeanPerier reviewed (approved):
Thank you for this clean-up Tom!

I only have one suggestion to make the DataLayout a std::uniq_ptr member to avoid starting a spread of its construction in helpers.

tblah committed:
Memoise construction of mlir::DataLayout

I decided to use std::optional instead of std::unique_ptr because we
don't need heap allocation here.
